server/aesdsocket.c

```diff
@@ -0,0 +1,331 @@
+    }
+
+    //Wait for data
+    int recv_ret;
+    //Start with a size, potentially incresing it if an entire packet cannot fit into it
+    recv_data = malloc(sizeof(char) * BUFFER_SIZE);
+    if(!recv_data){
+    	syslog(LOG_ERR, "Error occured during malloc = %s. Exiting...", strerror(errno));
+    	exit(-1);
+    }
+    int recv_idx = 0;
+    //How many 'BUFFER_SIZE' receive_blocks "recv_data" we have received from client till now
+    int receive_blocks = 1;
+    do {
+      recv_ret = recv(conn_fd, & recv_data[recv_idx], BUFFER_SIZE, 0);
+      if (recv_ret == -1) {
+        syslog(LOG_ERR, "Error occured while reading from socket = %s. Exiting...", strerror(errno));
+        exit(-1);
+      }
+//      printf("Received data = %s\n", &recv_data[recv_idx]);
+      recv_idx += recv_ret;
+
+      if (recv_idx != 0) {
+        if (recv_data[recv_idx - 1] == '\n') {
+
+          if (lseek(log_fd, 0, SEEK_END) == -1) {
+            syslog(LOG_ERR, "Error occured while lseek() to EOF = %s. Exiting...", strerror(errno));
+            exit(-1);
+          }
+
+          int write_byte_ret;
+          int write_len = recv_idx;
+          char * write_ptr = recv_data;
+          while (write_len != 0) {
+            write_byte_ret = write(log_fd, write_ptr, write_len);
+            
+            if (write_byte_ret == -1) {
+              if (errno == EINTR)
+                continue;
+              else {
+                syslog(LOG_ERR, "Error occured while writing to file = %s. Exiting...", strerror(errno));
+                exit(-1);
+              }
+            }
+            write_len -= write_byte_ret;
+            write_ptr += write_byte_ret;
+          }
+
+          file_size += recv_idx;
+
+          if (lseek(log_fd, 0, SEEK_SET) == -1) {
+            syslog(LOG_ERR, "Error occured while lseek() to BOF = %s. Exiting...", strerror(errno));
+            exit(-1);
+          }
+
+          char * log_read = malloc(sizeof(char) * file_size);
+          char * log_read_offset = log_read;
+          int bytes_to_read = file_size;
+          int bytes_read_count;
+          while (bytes_to_read != 0) {
+            bytes_read_count = read(log_fd, log_read_offset, bytes_to_read);
+            if (bytes_read_count == -1) {
+              if (errno == EINTR)
+                continue;
+              else {
+                syslog(LOG_ERR, "Error occured during reading from logfile = %s. Exiting...", strerror(errno));
+                exit(-1);
+              }
+
+            }
+            bytes_to_read -= bytes_read_count;
+            log_read_offset += bytes_read_count;
+          }
+          //Send the contents back to the client
+          send(conn_fd, log_read, file_size, 0);
+          //Free the used bufferr
+          free(log_read);
+
+          //Reset recv_idx to loop again
+          recv_idx = 0;
+        }
+        //Realloc the array if it got full without an '\n'
+        else if (recv_idx == (BUFFER_SIZE * receive_blocks)) {
```

Comment:
@prvbharadwaj1731 I don't think this logic will work correctly when recv_idx doesn't exactly equal BUFFER_SIZE * receive_blocks.  I think you will be exceeding array bounds in this case.  Try dropping your BUFFER_SIZE to a low value like 20 or 50 bytes and I think this code will fail.  I think larger packet sizes won't be handled properly.